doc/app/user_function.rst

```diff
@@ -1,102 +1,167 @@
------------------------
+.. note:: Check out the `cogent3 app cookiecutter template <https://github.com/cogent3/app_template>`_ that we've made available to make writing apps more accessible.
 
-If your function takes or returns ``cogent3`` types, you can use the existing type hints. To see what these are, use the ``defined_types()`` function.
+.. dropdown:: Specifying the type of app
 
-.. jupyter-execute::
+    The :ref:`supported app types <app_types>` are indicated by the ``AppType`` enum
+
+    .. jupyter-execute::
+
+        from cogent3.app.composable import AppType
+
+        list(AppType)
+
+    The decorator has a default value of ``"generic"``. This means that the app does data transformation and does not, for example, load data from disk or write data to disk (those are the ``loader`` and ``writer`` types).
+
+    If your application is not intended to be applied sequentially to a series of independent data records of the same type, then you set ``define_app(app_type=AppType.NON_COMPOSABLE)`` (or equivalently ``define_app(app_type="non_composable")``).
+    
+    .. note:: Non-composable apps cannot be added (or composed) with other apps.
+
+.. dropdown:: Handling not completed values
+
+    Cogent3 apps are designed to handle conditions under which data cannot be completely processed. For example, any errors during execution are stored in a :ref:`not completed <not_completed>` object (``NotCompleted``). All subsequent steps are skipped whenever an app returns a ``NotCompleted`` object.
+
+    These objects can also be used to stop the processing of a particular data record. For example, say you're writing an application that requires sequences to have some minimum length. If an input sequence is shorter than this, then you can create a :ref:`not completed <not_completed>` object and return it. This prevents any future processing, but the reasons for the failure (which you get to specify) can be saved for future reference.
+
+    You may want access to ``NotCompleted`` instances if, for example, you are developing an ``AppType.WRITER``. You can allow your code to "see" them with ``define_app(skip_not_completed=False)``.
 
-    from cogent3.app.typing import defined_types
+.. dropdown:: Supported cogent3 types
 
-    defined_types()
+    You can use the existing type hints if your function takes or returns ``cogent3`` types. To see these, use the ``defined_types()`` function.
 
-.. note:: You don't have to use cogent3 types, you can also use standard python types.
+    .. jupyter-execute::
 
-A simple example
-----------------
+        from cogent3.app.typing import defined_types
 
-Let's make an app that returns the elements of an alignment up to a specified index, with the index being a keyword argument. We now define a decorated function ``up_to`` and import the type hints and the decorator.
+        defined_types()
+
+    .. note:: You don't have to use cogent3 types. You can also use standard python types.
+
```

Comment:
Worth restating that apps can be defined from functions or from classes in your code, before describing both strategies.

Also I think there needs some guidance over where you might chose one over the other.  A function defined as an app is more terse, a class affords more nuance to discriminate between app initialization arguments and functional arguments.

Reply:
I don't want to overwhelm our potential developers with too much reading. So I'll skip this, although it's an important point.